### PR TITLE
LWDEV-4155 Pre-trade control

### PR DESCRIPTION
**Pull request by barsukovskii**
if resulting price differs from expected more than specified threshold - reject market order

### CODE REVIEW

File: src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt [hunk 1]
```diff
@@ -233,4 +243,16 @@ class MatchingEngine(private val LOGGER: Logger,
         val asset = if (order.isBuySide()) assetPair.quotingAssetId else assetPair.baseAssetId
         return balancesHolder.getAvailableBalance(order.clientId, asset)
     }
+
+    private fun checkPrice(order: NewOrder, price: Double, threshold: Double?): Boolean {
+        if (threshold == null) {
+            return true
+        }
+        val expectedPrice = order.expectedPrice() ?: return true
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt]: add 0 check for expected price - we don't have control over incoming data

File: src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt [hunk 2]
```diff
@@ -233,4 +243,16 @@ class MatchingEngine(private val LOGGER: Logger,
...
+        return if (order.isBuySide()) {
+            (price - expectedPrice) / expectedPrice < threshold
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt]: check case if threshold = 0.0

File: src/main/kotlin/com/lykke/matching/engine/services/MarketOrderService.kt [hunk 3]
```diff
@@ -93,13 +97,17 @@ class MarketOrderService(private val backOfficeDatabaseAccessor: BackOfficeDatab
...
 
-        val matchingResult = matchingEngine.match(order, orderBook)
+        val matchingResult = matchingEngine.match(order, orderBook, priceDifferenceThreshold = properties.priceDifferenceThreshold)
         val marketOrder = matchingResult.order as MarketOrder
         when (OrderStatus.valueOf(matchingResult.order.status)) {
             NoLiquidity -> {
                 rabbitSwapQueue.put(MarketOrderWithTrades(marketOrder))
                 writeResponse(messageWrapper, marketOrder, MessageStatus.NO_LIQUIDITY)
             }
+            PriceIsOutsideThreshold -> {
```
Comment (dmitrykrasovskih) [inline src/main/kotlin/com/lykke/matching/engine/services/MarketOrderService.kt]: add the same section to limit order services

File: src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt (hunk 1 above)
Comment (barsukovskii) [reply src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt]: added

File: src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt (hunk 2 above)
Comment (barsukovskii) [reply src/main/kotlin/com/lykke/matching/engine/matching/MatchingEngine.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/services/MarketOrderService.kt (hunk 3 above)
Comment (barsukovskii) [reply src/main/kotlin/com/lykke/matching/engine/services/MarketOrderService.kt]: added

**Comment by dmitrykrasovskih** [comment]
outdated

**Pull request closed without merging**